### ISSUE TITLE
Rescale weights (not gradients) column norms to 1

### ISSUE THREAD

**Issue by wchan** [tensorflow/tensorflow#608]
How is this possible in TensorFlow, I feel like it should be trivial, but I can't find anything in the API to do this =\

This is to rescale the columns of the weight matrix (i.e., input connections).

Thanks!

**Comment by nryant** [comment]
I implemented this used a slightly modified version of clip_by_norm, which I applied to each weights variable. I then computed these normed weights variables at each update and then ran a loop across the variables calling assign. Unfortunately, calling apply within a Python loop at each step was a major performance bottleneck and I can't see any op in the API that would allow me to move this update into the graph.

**Comment by vrv** [comment]
https://www.tensorflow.org/versions/master/api_docs/python/nn.html#l2_normalize , or an equivalent construct? (You should be able to compute the appropriate norm and scale using reductions / elementwise ops, right?)

**Comment by wchan** [comment]
My apologies, I didn't make it clear, what we usually do is we only normalize if the gradient column is > some threshold, i.e., if column norm > threshold, normalize, else do noop. so I dont think that Op will help, or am I mistaken?

**Comment by vrv** [comment]
There may be a better way to do it, but if I understand correctly, you may be able to use tf.select to do what you want (may not be the most efficient thing to do memory wise).

Assuming:
w_gradient is the matrix representing the gradients of shape [m, n]
w_gradient_column is whatever reduction function along the column dimension (shape [1, n])
w_tensor is the original weight matrix
w_scaled_tensor is w_tensor with the normalization applied for all columns

First you compute which columns to apply the rescaling to
condition_per_column = tf.greater(w_gradient_column, threshold)

Next you replicate that value along every row to make a boolean tensor of the same shape as w_tensor
condition_tiled = tf.tile(condition_per_column, [m, 1])

Lastly, you create a resulting tensor that applies a mask to select the right value from each tensor.
result = tf.select(condition_tiled, w_scaled_tensor, w_tensor)

Maybe this still isn't what you want, but hopefully you get the point: you can use tf.select and comparison operators to select/mask results conditionally.

**Comment by wchan** [comment]
I am going to write a TensorOp for this; in the TensorOp, is it legal to have the .Input() and the .Output() tensor point to the same memory? i.e., I want to do this in-place.

**Comment by wchan** [comment]
fyi, wrote a kernel for this:
https://github.com/wchan/tensorflow/blob/master/tensorflow/core/kernels/training_ops.cc

see: "ApplyMaxWeightColNorm"

code needs to be refactored, generalized and tests added before patch can be submitted "soon".

**Comment by girving** [comment]
I think a custom op for this is overkill.  Just do `matrix * tf.expand_dims(threshold / tf.maximum(threshold, norms), 1)`.  Hmm.  I guess for norms one has to do `tf.sqrt(tf.reduce_sum(tf.square(matrix), 1)` which isn't pretty, but if any new op is added here it should be an L2 norm op (or possibly some extra arguments the existing `L2Loss` op), not a normalization op.
